plugins/docker/stats_pumper.go

```diff
@@ -0,0 +1,393 @@
+	go func() {
+		// This will block until the container exits
+		err := client.Stats(docker.StatsOptions{
+			ID:    	id,
+			Stats:  statsrd,
+			Stream: true,
+			Done:   done,
+		})
+
+		// Once it has exited, close our pipes
+		close(statsrd)
+		close(done)
+		if err != nil {
+			close(done)
+			failure <- err
+		}
+	}()
+
+	container, err := m.client.InspectContainer(id)
+	if err != nil {
+		m.errors <- err
+	}
+	name := container.Name[1:]
+	fields, err := m.extractFields(id, m.client)
+	if m.nameFromEnv != "" {
+		if alt_name, ok := fields[m.nameFromEnv]; ok && alt_name != "" {
+			name = alt_name
+		}
+	}
+	fields["ContainerID"] = id
+	fields["ContainerName"] = name
+	var resultStats []*docker.Stats
+	for {
+		stats, ok := <-statsrd
+		if !ok {
+			break
+		}
+		resultStats = append(resultStats, stats)
+		m.Lock()
+		m.attached[id] = NewStatsPump(statsrd, name, fields)
+		m.Unlock()
+		m.send(&StatsAttachEvent{ID: id, Name: name, Type: "attach"})
+		return nil
+	}
+	return nil
+}
+
+// Inspect the container and extract the env vars/labels we were told to keep
+func (m *StatsManager) extractFields(id string, client DockerClient) (map[string]string, error) {
+
+	container, err := client.InspectContainer(id)
+	if err != nil {
+		return nil, err
+	}
+	name := container.Name[1:] // Strip the leading slas
+	image := container.Config.Image
+
+	m.ir.LogMessage("Container Name : " + name)
+	fields := getEnvVars(container, append(m.fieldsFromEnv, m.nameFromEnv))
+	if m.nameFromEnv != "" {
+		if alt_name, ok := fields[m.nameFromEnv]; ok && alt_name != "" {
+			name = alt_name
+		}
+	}
+	fields["ContainerID"] = id
+	fields["ContainerName"] = name
+	fields["ContainerImage"] = image
+
+	// NOTE! Anything that is a duplicate key will be overridden with the value
+	// that is in the label, not the environment
+	for _, key := range m.fieldsFromLabels {
+		if value, ok := container.Config.Labels[key]; ok {
+			fields[key] = value
+		}
+	}
+
+	return fields, nil
+}
+
+func (m *StatsManager) send(event *StatsAttachEvent) {
+	m.RLock()
+	defer m.RUnlock()
+	for ch, _ := range m.channels {
+		ch <- event
+	}
+}
+
+func (m *StatsManager) addStatsListener(ch chan *StatsAttachEvent) {
+	m.ir.LogMessage(" addListener ")
+	m.Lock()
+	defer m.Unlock()
+	m.channels[ch] = struct{}{}
+	go func() {
+		for id, pump := range m.attached {
+			m.ir.LogMessage("StatsID:" + id +" ,  Name: "+  pump.Name +", Type:  attach")
+			ch <- &StatsAttachEvent{ID: id, Name: pump.Name, Type: "attach"}
+		}
+	}()
+}
+
+func (m *StatsManager) removeStatsListener(ch chan *StatsAttachEvent) {
+	m.Lock()
+	defer m.Unlock()
+	delete(m.channels, ch)
+}
+
+func (m *StatsManager) Get(id string) *StatsPump {
+	m.Lock()
+	defer m.Unlock()
+	return m.attached[id]
+}
+
+func (s *Source) AllContainers() bool {
+	return s.ID == "" && s.Name == "" && s.Filter == ""
+}
+
+// Try to reboot all of our connections
+func (m *StatsManager) restart() error {
```

Comment:
Yeah, there's getting to be enough shared code here that some of this should be just another file with those shared methods in it. When that gets unwieldy then it could eventually be broken out more.